gen/enum.go

```diff
@@ -101,6 +106,99 @@ func enum(g Generator, spec *compile.EnumSpec) error {
 			<end>
 			return fmt.Sprintf("<$enumName>(%d)", <$w>)
 		}
+
+		func (<$v> <$enumName>) MarshalJSON() ([]byte, error) {
+			<if len .Spec.Items>
+				<$w> := int32(<$v>)
+				switch <$w> {
+				<range .UniqueItems>
+					case <.Value>:
+						return ([]byte)("\"<.Name>\""), nil
+				<end>
+				}
+			<end>
+			return ([]byte)(<$strconv>.FormatInt(int64(<$v>), 10)), nil
+		}
+
+		func (<$v> *<$enumName>) UnmarshalJSON(text []byte) error {
+			<$d := newVar "d">
+			<$t := newVar "t">
+			<$err := newVar "err">
+
+			<$d> := <$json>.NewDecoder(<$bytes>.NewReader(text))
+			<$d>.UseNumber()
+			<$t>, <$err> := <$d>.Token()
+
+			if <$err> != nil {
+				return <$err>
+			}
+
+			<$ok := newVar "ok">
+			if <$w>, <$ok> := <$t>.(<$json>.Number); <$ok> {
+				<$x := newVar "x">
+				<$x>, <$err> := <$w>.Int64()
+				if <$err> != nil {
+					return <$err>
+				}
+				if <$x> <">="> 0x80000000 {
+					return <$fmt>.Errorf("enum overflow from JSON %q for %q", text, "<$enumName>")
```

Comment:
Don't need to format `<$enumName>` using `Errorf`. It's known at codegen time so you can make it part of the string.